Fixed #25181 -- Added localdate() function to get date in a different time zone.

jdufresne opened the pull request:
Thanks @kswiat for the original [PR](https://github.com/django/django/pull/5055).

<https://code.djangoproject.com/ticket/25181>

Some response to other tickets and PRs:

From @aaugustin 

> It may be useful to make these functions fallback reasonably when `USE_TZ` is `False` for the benefit of pluggable applications. In that case, `localtime()` should return:
> 
> ...
> 3. (probably) an exception if `value` is provided and it is aware

Adding the following guard to `localtime()` produces a [few test failures](http://dpaste.com/2S6RD55). It looks like there is an existing assumption that `localtime()` works on aware datetimes regardless of the value `USE_TZ`. I'm open towards moving in the suggested direction, but it may be more invasive than originally thought and have backwards compatibility concerns. What are you thoughts?

Guard: 

```python
if not settings.USE_TZ and is_aware(value):
    raise ValueError("Cannot call localtime() with an aware datetime when USE_TZ is False")
```

---

If agreeable, I would also like to add an analogous function to return `datetime.time()` in the current time zone. (Obviously, the name `localtime()` won't work.) I think it would be just as useful as `localdate()` for the same reasons. IME, just the current TZ's time is frequently  important to display to the user or compare to other values. Having an official, documented convenience function that gets this right and is reusable across applications seems beneficial to me. However, @aaugustin didn't give it the same "+0" as `localdate()` in ticket [#27082](https://code.djangoproject.com/ticket/27082). So I'm looking for feedback.

Thanks!

aaugustin commented:
Disclaimer: I'm fear that I'm not doing a good job at providing consistent feedback on this issue, apologies for that.

Generally speaking, whenever possible, it's best for `django.utils.*` not to depend on settings. That makes utils modules less coupled to Django and easier to test. I'm not sure if we can in this particular case, but it's a design principle to keep in mind.

Originally `localtime()` was a conversion function designed to receive an aware datetime and return an aware datetime. (As I explained on the ticket it's primarily a wrapper around the pattern that pytz requires to perform that conversion.) This operation is legal regardless of the value of `settings.USE_TZ`.

----

It's unclear to me that this part is a good idea:

```python
    if not settings.USE_TZ and is_naive(value):
        return value
```

(Sorry if I'm the person who suggested it.)

As far as I can tell, the intent is to make `localtime()` Just Work when `value is None and not settings.USE_TZ`, but this behavior may be considered surprising.

Here's an attempt at defining a consistent behavior for `localtime()`:

`USE_TZ` | `value` | `timezone` | behavior
-------- | ------- | ---------- | --------
`True`   |  `None` |     `None` | return aware now in current time zone — obvious
`True`   |  `None` |   timezone | return aware now in specified time zone — obvious
`True`   |   naive |     `None` | assume value is in the default time zone, raise a warning, convert it to the current time zone, return an aware value — fairly obvious: converting from the default time zone with a warning is Django's standard behavior when it receives a naive datetime in a context that requires an aware datetime (note that currently Python ≥ 3.6 silently assumes the default time zone and Python < 3.6 raises an exception; not raising an exception is a backwards incompatible change, but not an awful one)
`True`   |   naive |   timezone | assume value is in the default time zone, raise a warning, convert it to the specified time zone, return an aware value — fairly obvious: as above (same comment as above)
`True`   |   aware |     `None` | convert aware value to current time zone — obvious
`True`   |   aware |   timezone | convert aware value to specified time zone — obvious
`False`  |  `None` |     `None` | return 1. aware now OR 2. naive now in current time zone
`False`  |  `None` |   timezone | return 1. aware now OR 2. naive now in specified time zone
`False`  |   naive |     `None` | assume value is in the default time zone, raise a warning, convert it to the current time zone, return an 1. aware OR 2. naive value
`False`  |   naive |   timezone | assume value is in the default time zone, raise a warning, convert it to the current time zone, return an 1. aware OR 2. naive value
`False`  |   aware |     `None` | convert aware value to current time zone — almost obvious:  `localtime`'s historical definition as a conversion function for aware datetimes, regardless of `settings.USE_TZ`, mandates this behavior
`False`  |   aware |   timezone | convert aware value to specified time zone — almost obvious: as above

The only point I'm unsure about is whether `localtime` should 1. always return an aware datetime? or 2. return a naive datetime when `USE_TZ = False` and `value` is naive (which includes the case where `value` is not provided, as it will default to `now()`, a naive value).

Having a consistent return type i.e.`localtime` always returns an aware datetime is valuable. However one may argue that returning a naive datetime when `USE_TZ = False` and `value` is naive makes it comparable with other values manipulated by Django.

I'm leaning towards the first solution, always returning an aware datetime, for two reasons:

- this is easier to explain and to document (see also the principle of least astonishment)
- `localtime()` is primarily useful for communicating with external systems that only know about local time. If you're displaying datetimes in templates or storing them in the database, Django already does the right thing, you don't need to convert yourself.

The second solution feels to me like we're trying to cram too much functionality in a single function.

aaugustin commented:
> If agreeable, I would also like to add an analogous function to return datetime.time() in the current time zone. (Obviously, the name localtime() won't work.) I think it would be just as useful as localdate() for the same reasons. IME, just the current TZ's time is frequently important to display to the user or compare to other values. Having an official, documented convenience function that gets this right and is reusable across applications seems beneficial to me. However, @aaugustin didn't give it the same "+0" as localdate() in ticket #27082. So I'm looking for feedback.

A function that extracts the time part feels less useful to me, but it doesn't hurt either, so go 
ahead if you want it.

The hardest part is to figure out a consistent naming scheme before moving forwards. I'm aware (ha, ha) that the original choice of `localtime()` doesn't help.

If we can't come up with a good enough naming scheme, `localtime().date()` and `localtime().time()` don't look too bad to me. Once we've made it possible to call `localtime()` without arguments, we could get away with documenting these patterns.

For the avoidance of doubt: you don't need any of this to display dates or times to users with Django templates. The template engine does the right thing.

jdufresne commented:
> Generally speaking, whenever possible, it's best for django.utils.* not to depend on settings. That makes utils modules less coupled to Django and easier to test. I'm not sure if we can in this particular case, but it's a design principle to keep in mind.

This makes a lot of sense to me. I think we can improve the current `localtime()`/`localdate()` API without introducing this dependency:

* Keep `timezone_today()` function where as it currently is
* Change `localtime()` to default to `timezone.now()`
* Introduce `localdate()`; implemented as `localtime().date()`

This adds a lot of convenience for projects with `USE_TZ=True` as now there is s a single call for the active time zone's today value and a single call with no arguments for the active time zone's now value. I think, for now, `timezone.localtime().time()` is simple enough, I can certainly document the pattern.

For people with `USE_TZ=False`, they can continue to use Python's `datetime.date.today()`.

What do you think about this? I have implemented it in the most recent change, I will squash commits if it looks like a good direction. I think it keeps this change simple. Otherwise I will implement the value table you've outlined above.

aaugustin commented:
@jdufresne I'm OK with the current version patch. Keeping the documented restriction that `localtime` doesn't work on naive datetimes indeed minimizes the impact of this change. Thanks.

aaugustin commented:
Go ahead and merge it if you're happy with it as well :-)

jdufresne commented:
Thanks. Squashed the commits. 

With this change in direction, do you feel it is still correct to mark the ticket as "Fixed" or should this be downgraded to "Refs"?

timgraham commented on docs/ref/utils.txt:
```diff
@@ -974,14 +974,32 @@ appropriate entities.
 
+    .. versionchanged:: 1.11
+
+        The ``value`` argument now defaults to :func:`now`.
```
To make it clearer about the old behavior, I'd instead instead say, "In older versions, `value` is a required argument."

timgraham commented on docs/ref/utils.txt:
```diff
@@ -974,14 +974,32 @@ appropriate entities.
+
+    Converts an aware :class:`~datetime.datetime` to a different time zone, by
+    default the :ref:`current time zone <default-current-time-zone>`, using
+    :func:`localtime` and returns :meth:`~datetime.datetime.date`.
```
"returns a date"?

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -325,6 +325,12 @@ URLs
 
 * ...
 
+Utilities
```
In past releases, we haven't included changes in django.utils in the release notes.

timgraham commented on docs/releases/1.11.txt:
```diff
@@ -325,6 +325,12 @@ URLs
+~~~~~~~~~
+
+* The new :func:`django.utils.timezone.localdate` function returns a
+  :class:`~datetime.datetime`'s date in the current time zone.
```
You would need to use a right quote here to prevent Sphinx from rendering it in the opposite direction.

timgraham commented on tests/utils_tests/test_timezone.py:
```diff
@@ -54,6 +57,24 @@ def test_now(self):
 
+    def test_localdate(self):
+        naive = datetime.datetime(2015, 1, 1, 0, 0, 1)
+        with self.assertRaisesMessage(ValueError, 'astimezone() cannot be applied to a naive datetime'):
```
Please check with Python 3.6, no exception is raised here (see a7a7ecd2b026c61a39a46d2d7eced0e06a92c970).

aaugustin replied:
See also my recent Trac ticket suggesting to normalize behavior across Python versions.

jdufresne replied:
[Ticket #27138](https://code.djangoproject.com/ticket/27138).

Unless you think that ticket should be solved first, I'll follow the existing pattern found in the other tests for this PR.

timgraham commented on tests/utils_tests/test_timezone.py:
```diff
@@ -54,6 +57,24 @@ def test_now(self):
+        with timezone.override(EAT):
+            self.assertEqual(timezone.localdate(aware), datetime.date(2014, 12, 31))
+
+        with mock.patch('django.utils.timezone.now') as mock_now:
```
I think this works fine and is a bit simpler: `with mock.patch('django.utils.timezone.now', return_value=aware):`

timgraham commented on docs/ref/utils.txt:
```diff
@@ -974,14 +974,32 @@ appropriate entities.
+    .. versionadded:: 1.11
+
+    Converts an aware :class:`~datetime.datetime` to a different time zone, by
+    default the :ref:`current time zone <default-current-time-zone>`, using
```
I tried to reword this sentence, see what you think.

Uses :func:`localtime` to convert an aware :class:`~datetime.datetime` to a :meth:`~datetime.datetime.date` in a different time zone, by default the :ref:`current time zone <default-current-time-zone>`.

jdufresne commented:
Made recommended edits, thanks.

There are some Jenkins failures, but they look unrelated to the changes.

timgraham commented on django/utils/timezone.py:
```diff
@@ -299,13 +299,18 @@ def template_localtime(value, use_tz=None):
+    Only aware datetimes are allowed. When value is omitted, it defaults to
+    now().
+
+    Local time is defined by the current time zone, unless another time zone is
```
I wouldn't have reflowed this line since you didn't make any other changes and it would simplify the diff.

timgraham commented:
I think we can open a new ticket if there's any other related changes to make. I retitled the ticket.
